### PR TITLE
Add Cache-Side Config Generator

### PR DESCRIPTION
**Pull request by rob05c**
## What does this PR (Pull Request) do?

Adds a cache-side config generator. This is a small go binary (`atstccfg`), which ORT calls instead of requesting Traffic Ops. The `atstccfg` generator then generates configs it can, and proxies to TO for configs it can't yet generate.

This PR includes generating `parent.config` and all `profile` configs, with everything else being proxied. The plan is to move configs over one at a time, until all configs are generated cache-side.

See https://cwiki.apache.org/confluence/display/TC/Cache-Side+Config+Generation

It also moves the logic into `/lib/go-atscfg`, a library that can be used if someone wants to generate configs themselves somewhere else.

I've parity-tested the generated configs, against Traffic Ops generation, on all our production edges, mids, and profiles. They all diff identical (except where order is nondeterministic in Perl).

Includes documentation, explaining the new config generation, what's generated where and how.
Includes unit tests.
Includes Changelog.

- [x] This PR is not related to any other Issue

## Which Traffic Control components are affected by this PR?

- Traffic Ops
- Traffic Ops ORT

## What is the best way to verify this PR?

Install ORT, run, verify config files match previous versions / TO.

## If this is a bug fix, what versions of Traffic Control are affected?

Not a bug fix.

## The following criteria are ALL met by this PR

- [x] This PR includes tests OR I have explained why tests are unnecessary
- [x] This PR includes documentation OR I have explained why documentation is unnecessary
- [x] This PR includes an update to CHANGELOG.md OR such an update is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR ensures that database migration sequence is correct OR this PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY** (see [the Apache Software Foundation's security guidelines](https://www.apache.org/security/) for details)

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/4052/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/4107/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/4108/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/4110/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/4111/
Test FAILed.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/4112/
Test FAILed.

File: docs/source/admin/traffic_server.rst [hunk 1]
```diff
@@ -80,6 +80,21 @@ Configuring Traffic Server
 ==========================
 All of the :abbr:`ATS (Apache Traffic Server)` application configuration files are generated by Traffic Ops and installed by :term:`ORT`. The :file:`traffic_ops_ort.pl` file should be installed on all :term:`cache server` s (See `Installing the ORT Script`_), usually in ``/opt/ort``. It is used to do the initial install of the configuration files when the :term:`cache server` is being deployed, and to keep the configuration files up-to-date when the :term:`cache server` is already in service.
 
+.. _config-generation:
+
+ORT Config File Generation
+-------------------------
+
+In the past, ATS config files were generated by Traffic Ops. Traffic Control is in the process of moving ATS config file generation, to a library for generic use, and to an application which uses the library and resides on the cache.
```
Comment (ezelkow1) [inline docs/source/admin/traffic_server.rst]: remove comma after generation

**Comment by rob05c** [comment]
Addressed PR comments, fixed merge conflicts.

**Comment by rob05c** [comment]
retest this please

**Comment by ocket8888** [comment]
License header problems:

```
Error                                 Unknown! traffic_ops/ort/atstccfg/urisigningconfig_test.go
Error                                 Unknown! traffic_ops/ort/atstccfg/urlsigconfig_test.go
```

**Comment by ocket8888** [comment]
It also doesn't like the license of your `pkg` dependency:

```
Error                                    BSD~! vendor/github.com/pkg/errors/.travis.yml
Error                                     BSD! vendor/github.com/pkg/errors/LICENSE
Error                                    BSD~! vendor/github.com/pkg/errors/Makefile
Error                                    BSD~! vendor/github.com/pkg/errors/appveyor.yml
Error                                    BSD~! vendor/github.com/pkg/errors/bench_test.go
Error                                    BSD~! vendor/github.com/pkg/errors/errors.go
Error                                    BSD~! vendor/github.com/pkg/errors/errors_test.go
Error                                    BSD~! vendor/github.com/pkg/errors/example_test.go
Error                                    BSD~! vendor/github.com/pkg/errors/format_test.go
Error                                    BSD~! vendor/github.com/pkg/errors/json_test.go
Error                                    BSD~! vendor/github.com/pkg/errors/stack.go
Error                                    BSD~! vendor/github.com/pkg/errors/stack_test.go
```

File: LICENSE [hunk 2]
```diff
@@ -440,3 +440,7 @@ For the lestrrat-go/jwx (commit e35178a) component:
 For the dgrijalva/jwt-go (commit 5e25c22) component:
 @traffic_ops/traffic_ops_golang/vendor/github.com/dgrijalva/jwt-go/*
 ./traffic_ops/traffic_ops_golang/vendor/github.com/dgrijalva/jwt-go/LICENSE
+
+For the ogier/pflag component:
```
Comment (rawlinp) [inline LICENSE]: For any new vendored dependencies please include the commit hash of when they were cloned into the repo.

Comment (ocket8888) [reply LICENSE]: And/Or semver tag, if available

**Comment by rawlinp** [comment]
Where did we land on the idea to share code between the new atstccfg tool and the API and just replacing DB queries w/ API calls? Are we doing that?

**Comment by rob05c** [comment]
>License header problems
Fixed.

File: LICENSE (hunk 2 above)
Comment (rob05c) [reply LICENSE]: Done.

**Comment by rob05c** [comment]
>Where did we land on the idea to share code between the new atstccfg tool and the API and just replacing DB queries w/ API calls? Are we doing that?

All the logic is shared, in `/lib/go-atscfg`. The data loading is still using regular DB calls, though. I didn't see a clean way to use the API within TO. It's technically possible, but it would be much slower, and load lots of unnecessary data.

**Comment by rawlinp** [comment]
> All the logic is shared, in /lib/go-atscfg. The data loading is still using regular DB calls, though. I didn't see a clean way to use the API within TO. It's technically possible, but it would be much slower, and load lots of unnecessary data.

Yeah, I don't think it's necessary to try to use the API within TO. Thanks!

**Pull request merged**